db.go

```diff
@@ -1371,3 +1388,49 @@ func (db *DB) DropPrefix(prefix []byte) error {
 	db.opt.Infof("DropPrefix done")
 	return nil
 }
+
+// Subscribe can be used watch key changes for the given key prefix
+func (db *DB) Subscribe(ctx context.Context, cb callback, prefix []byte, prefixes ...[]byte) error {
+	if cb == nil {
+		return ErrNilCallback
+	}
+	prefixes = append(prefixes, prefix)
+	c := y.NewCloser(1)
+	recvCh, id := db.pub.newSubscriber(c, prefixes...)
+	slurp := func(batch *pb.KVList) {
+		defer func() {
+			if len(batch.GetKv()) > 0 {
+				cb(batch)
+			}
+		}()
+		for {
+			select {
+			case kvs := <-recvCh:
```

Comment:
My previous question still stands whether draining would cause callback to never get executed. You said that there is a cap in doWrites. Please point me to that.

Reply:
the cap is there for request batching so I removed the draining part in publisher goroutine  https://github.com/dgraph-io/badger/blob/master/db.go#L694 if all the DB writes points to the same prefix then there may be a chance of callback never get executed

Reply:
I think this only limits the size of each element in the channel, not the size of the channel. In certain high load scenarios, it is possible that channel never becomes empty and we keep increasing the size of the batch while draining (this step). Let's take a step back and help me understand the point of doing the batching of `reqs`, why not call `cb` on one `reqs`?

Reply:
@mangalaman93 -- would be worth reading this article: https://mechanical-sympathy.blogspot.com/2011/10/smart-batching.html

Reply:
But aren't we already doing batching, because we have group of `req` coming in? And, if I am missing something and this batching is still useful, we should have a similar cap on the size of batching as the one Balaji points to.